File: src/Tests/Behavioral/DeferTypelessReturns/DeferTypelessReturns.cs
namespace go;

using fmt = fmt_package;

partial class main_package {

[GoType] partial struct item {
    internal nint n;
}

[GoType("num:uintptr")] partial struct handle;

internal static readonly handle invalid = /* ^handle(0) */ unchecked((handle)18446744073709551615);

internal static readonly UntypedInt appErr = /* 1 << 29 */ 536870912;

internal static readonly handle big = /* appErr + 5 */ 536870917;

internal static nint idx(this handle h) {
    return (nint)(uintptr)(h - (handle)appErr);
}

internal static (ж<item>, error) find(slice<item> xs, nint want) => func<(ж<item>, error)>((defer, recover) => {
    deferǃ(ᴛ1 => fmt.Println(ᴛ1), "find done", defer);
    foreach (var (i, _) in xs) {
        if (xs[i].n == want) {
            return (Ꮡ(xs, i), default!);
        }
    }
    return (default!, fmt.Errorf("not found"u8));
});

internal static error closeIt(ж<nint> Ꮡp, nint tag) {
    ref var p = ref Ꮡp.DerefOrNil();

    fmt.Println("closeIt", Ꮡp == nil, tag);
    return default!;
}

internal static byte first(array<byte> value) {
    value = value.Clone();

    var p = Ꮡ(value).at<byte>(0);
    return p.Value;
}

internal static void Main() => func((defer, recover) => {
    deferǃ(closeIt, (ж<nint>)(nil), (nint)(3), defer);
    ref var h = ref heap<res>(out var Ꮡh);
    h = new res(id: 4);
    var hʗ1 = h;
    defer(() => hʗ1.close());
    var xs = new item[]{new(1), new(2)}.slice();
    var (p, err) = find(xs, 2);
    fmt.Println(p != nil, err == default!);
    (_, err) = find(xs, 9);
    fmt.Println(err);
    fmt.Println(first(new byte[]{7, 8, 9, 10}.array()));
    fmt.Println(big.idx());
    fmt.Println(invalid == big, invalid == invalid);
});

[GoType] partial struct res {
    internal nint id;
}

internal static error close(this res h) {
    fmt.Println("closed", h.id);
    return default!;
}

} // end main_package

File: src/Tests/Behavioral/DeferTypelessReturns/DeferTypelessReturns.go
package main

import "fmt"

type item struct{ n int }
type handle uintptr

const invalid = ^handle(0)        // typed named-uintptr const, folded value > int32 (CS0266 shape)
const appErr = 1 << 29            // untyped const
const big handle = appErr + 5     // named-uintptr const, folded beyond int32 (same CS0266 shape)

// idx subtracts an untyped const from a named-uintptr receiver (CS0034 shape: the cast on the
// const must target the NAMED type, not golib's uintptr struct, to stay unambiguous).
func (h handle) idx() int { return int(h - appErr) }

// find has unnamed results, a defer, and EVERY return contains nil — no return statement
// contributes a natural type, so the value-returning execution wrapper needs its explicit
// result type argument (CS8030 shape: syscall getProcessEntry).
func find(xs []item, want int) (*item, error) {
	defer fmt.Println("find done")
	for i := range xs {
		if xs[i].n == want {
			return &xs[i], nil
		}
	}
	return nil, fmt.Errorf("not found")
}

// closeIt is a deferred method-value call passing untyped nil to a pointer param (CS0123 shape:
// the nil argument must cast to the parameter's C# type so deferǃ's generic inference binds).
func closeIt(p *int, tag int) error {
	fmt.Println("closeIt", p == nil, tag)
	return nil
}

// first takes the address of an element of a BY-VALUE ARRAY PARAMETER (CS0103 shape: params are
// cloned but never heap-boxed, so the element address boxes a copy sharing element storage).
func first(value [4]byte) byte {
	p := &value[0]
	return *p
}

func main() {
	defer closeIt(nil, 3)
	// A no-arg deferred METHOD whose signature RETURNS a value (registry Key.Close in
	// time's zoneinfo_windows): the bare method-group form is a Func<error> (CS1503) -
	// the lambda form discards the result, matching Go deferred-call semantics.
	h := res{id: 4}
	defer h.close()
	xs := []item{{1}, {2}}
	p, err := find(xs, 2)
	fmt.Println(p != nil, err == nil)
	_, err = find(xs, 9)
	fmt.Println(err)
	fmt.Println(first([4]byte{7, 8, 9, 10}))
	fmt.Println(big.idx())
	fmt.Println(invalid == big, invalid == invalid)
}

// res.close returns error - the defer must not bind it as a bare method group.
type res struct{ id int }

func (h res) close() error {
	fmt.Println("closed", h.id)
	return nil
}
